server/src/main/scala/za/co/absa/atum/server/api/http/Server.scala

```diff
@@ -0,0 +1,101 @@
+import org.http4s.blaze.server.BlazeServerBuilder
+import org.http4s.server.Router
+import sttp.monad.MonadError
+import sttp.tapir.generic.auto.schemaForCaseClass
+import sttp.tapir.json.play.jsonBody
+import sttp.tapir.server.http4s.Http4sServerOptions
+import sttp.tapir.server.http4s.ztapir.ZHttp4sServerInterpreter
+import sttp.tapir.server.interceptor.DecodeFailureContext
+import sttp.tapir.server.interceptor.decodefailure.DecodeFailureHandler
+import sttp.tapir.server.interceptor.decodefailure.DefaultDecodeFailureHandler.respond
+import sttp.tapir.server.model.ValuedEndpointOutput
+import sttp.tapir.swagger.bundle.SwaggerInterpreter
+import sttp.tapir.ztapir._
+import sttp.tapir.{DecodeResult, PublicEndpoint, headers, statusCode}
+import za.co.absa.atum.server.Constants.{SwaggerApiName, SwaggerApiVersion}
+import za.co.absa.atum.server.api.controller._
+import za.co.absa.atum.server.model.BadRequestResponse
+import zio.interop.catz._
+import zio.{RIO, ZIO}
+
+trait Server extends Endpoints {
+
+  type Env = PartitioningController with CheckpointController
+  type F[A] = RIO[Env, A]
+
+  private val decodeFailureHandler: DecodeFailureHandler[F] = new DecodeFailureHandler[F] {
+    override def apply(ctx: DecodeFailureContext)(implicit monad: MonadError[F]): F[Option[ValuedEndpointOutput[_]]] = {
+      monad.unit(
+        respond(ctx).map { case (sc, hs) =>
+          val message = ctx.failure match {
+            case DecodeResult.Missing => s"Decoding error - missing value."
+            case DecodeResult.Multiple(vs) => s"Decoding error - $vs."
+            case DecodeResult.Error(original, _) => s"Decoding error for an input value '$original'."
+            case DecodeResult.Mismatch(_, actual) => s"Unexpected value '$actual'."
+            case DecodeResult.InvalidValue(errors) => s"Validation error - $errors."
+          }
+          val errorResponse = BadRequestResponse(message)
+          ValuedEndpointOutput(statusCode.and(headers).and(jsonBody[BadRequestResponse]), (sc, hs, errorResponse))
+        }
+      )
+    }
+  }
+
+  private val http4sServerOptions: Http4sServerOptions[F] = Http4sServerOptions
+    .customiseInterceptors[F]
+    .decodeFailureHandler(decodeFailureHandler)
+    .options
+
+  private def createServerEndpoint[I, E, O](
+    endpoint: PublicEndpoint[I, E, O, Any],
+    logic: I => ZIO[Env, E, O]
+  ): ZServerEndpoint[Env, Any] = {
+    endpoint.zServerLogic(logic).widen[Env]
+  }
+
+  private def createAllServerRoutes: HttpRoutes[F] = {
+    val endpoints = List(
+      createServerEndpoint(createCheckpointEndpoint, CheckpointController.createCheckpoint),
+      createServerEndpoint(createPartitioningEndpoint, PartitioningController.createPartitioningIfNotExists)
+    )
+    ZHttp4sServerInterpreter[Env](http4sServerOptions).from(endpoints).toRoutes
+  }
+
+  private def createSwaggerRoutes: HttpRoutes[F] = {
+    val endpoints = List(createCheckpointEndpoint, createPartitioningEndpoint)
```

Comment:
perhaps consider extracting these elsewhere so that it's automatically loaded. I.e., whoever adds more endpoints must add such newly created classes/objects to multiple places, and here I propose to implement it in a way that keep such places to minimum

Reply:
At this point I can't think of a way how to reduce it. `createAllServerRoutes` creates Http4s server routes and you need to map each endpoint to its controller.